packages/public-api/src/search/searchService.ts

```diff
@@ -0,0 +1,44 @@
+import { getRepository } from 'typeorm';
+import { Hostname, PackagePopularityView } from '@gradejs-public/shared';
+
+const ESCAPED_SEARCH_SYMBOLS_REGEXP = /[%_]/g;
+const SEARCH_RESULT_PER_ENTITY_LIMIT = 3;
+
+export async function searchEntitiesByName(namePartial: string) {
+  const filteredNamePartial = namePartial
+    .toLowerCase()
+    .replace(ESCAPED_SEARCH_SYMBOLS_REGEXP, (match) => `\\${match}`);
+
+  if (!filteredNamePartial.length) {
+    return { hostnames: [], packages: [] };
+  }
+
+  const packagePopularityViewRepo = getRepository(PackagePopularityView);
+  const hostnameRepo = getRepository(Hostname);
+
+  const packageSearchQuery = packagePopularityViewRepo
+    .createQueryBuilder('package')
+    .select('package.packageName')
+    .where('package.packageName like :namePartial', { namePartial: `%${filteredNamePartial}%` })
+    .orderBy('package.usageByHostnameCount', 'DESC')
+    .limit(SEARCH_RESULT_PER_ENTITY_LIMIT)
+    .getMany();
+
+  const hostnameSearchQuery = hostnameRepo
+    .createQueryBuilder('hostname')
+    .select('hostname.hostname')
```

Comment:
`.join(WebPage, "webpage", "webpage.hostname_id = hostname.id")` + join `WebPageScan` and `getRawMany`

Reply:
This will return all of the scans for each hostname with no way to limit them. I'll add a proper subquery later in.